packages/material-ui/src/Unstable_TrapFocus/Unstable_TrapFocus.js

```diff
@@ -16,9 +26,11 @@ function Unstable_TrapFocus(props) {
     disableRestoreFocus = false,
     getDoc,
     isEnabled,
+    focusSelectors = [],
     open,
   } = props;
   const ignoreNextEnforceFocus = React.useRef();
+  const lastEvent = React.useRef(null);
```

Comment:
I don't think that we need this variable, what about?

```diff
diff --git a/packages/material-ui/src/Unstable_TrapFocus/Unstable_TrapFocus.js b/packages/material-ui/src/Unstable_TrapFocus/Unstable_TrapFocus.js
index 603eac26f..6ce1aba39 100644
--- a/packages/material-ui/src/Unstable_TrapFocus/Unstable_TrapFocus.js
+++ b/packages/material-ui/src/Unstable_TrapFocus/Unstable_TrapFocus.js
@@ -30,7 +30,6 @@ function Unstable_TrapFocus(props) {
     open,
   } = props;
   const ignoreNextEnforceFocus = React.useRef();
-  const lastEvent = React.useRef(null);
   const sentinelStart = React.useRef(null);
   const sentinelEnd = React.useRef(null);
   const nodeToRestore = React.useRef();
@@ -59,9 +58,8 @@ function Unstable_TrapFocus(props) {
     nodeToRestore.current = getDoc().activeElement;
   }

-  const onSentinelFocus = React.useCallback(() => {
+  const onSentinelFocus = React.useCallback((position) => () => {
     const focusRadios = rootRef.current.querySelectorAll('input:checked');
-    const isShiftTab = Boolean(lastEvent.current?.shiftKey && lastEvent.current?.keyCode === 9);
     const selectors = [...focusSelectorsRoot, ...focusSelectors].filter(Boolean);
     const focusChildren = rootRef.current.querySelectorAll(selectors.join(', '));
     const isFocusChildren = Boolean(focusChildren?.length);
@@ -71,7 +69,7 @@ function Unstable_TrapFocus(props) {
       focusChildren[focusChildren.length - 1].type === 'radio'
         ? focusRadios[0]
         : focusChildren[focusChildren.length - 1];
-    if (isShiftTab) {
+    if (position === 'start') {
       return focusEnd.focus();
     }
     return focusStart.focus();
@@ -142,7 +140,6 @@ function Unstable_TrapFocus(props) {
     };

     const loopFocus = (nativeEvent) => {
-      lastEvent.current = nativeEvent;
       // 9 = Tab
       if (disableEnforceFocus || !isEnabled() || nativeEvent.keyCode !== 9) {
         return;
@@ -209,9 +206,9 @@ function Unstable_TrapFocus(props) {

   return (
     <React.Fragment>
-      <div onFocus={onSentinelFocus} tabIndex={0} ref={sentinelStart} data-test="sentinelStart" />
+      <div onFocus={onSentinelFocus('start')} tabIndex={0} ref={sentinelStart} data-test="sentinelStart" />
       {React.cloneElement(children, { ref: handleRef, onFocus })}
-      <div onFocus={onSentinelFocus} tabIndex={0} ref={sentinelEnd} data-test="sentinelEnd" />
+      <div onFocus={onSentinelFocus('end')} tabIndex={0} ref={sentinelEnd} data-test="sentinelEnd" />
     </React.Fragment>
   );
 }
```

It fails some of the tests which surface how we could better write them :)

Reply:
Nicely done!  I'll make the update.